_posts/2019-12-11-reproducible-firmware-builds.md

```diff
@@ -0,0 +1,616 @@
+---
+
+If you have ever worked on a large scale embedded project before, you've probably run into situations where the build system or binary behaves differently depending on where it was compiled. For example, maybe the binary completely fails to compile on one computer and on another it compiles but crashes on boot!
+
+<!-- excerpt start -->
+
+In this article we will discuss what a **Reproducible Build** is, walk through the process of updating a firmware project so the build _is_ reproducible, and explore how we can leverage what we have learned for other aspects of development.
+
+<!-- excerpt end -->
+
+{:.no_toc}
+
+## Table of Contents
+
+<!-- prettier-ignore -->
+* auto-gen TOC:
+{:toc}
+
+## Reproducible Builds
+
+A build is said to be "Reproducible" when, given a set of source code, the exact same binary can be generated by different people on different computers. Specifically, this means:
+
+- code size and placement of sections are the same
+- filepaths emitted in the binary are the same
+- when a binary diff is computed between the build artifacts (e.g the ELF or .bin), there will be zero differences
+
+Many open source projects (such as Debian GNU/Linux) are working towards making the entire build reproducible[^8].
+
+### Why does it matter if the build is Reproducible?
+
+A common problem that arises when a build is not reproducible is the binary behaving differently for developers on the team. Debugging and supporting these kinds of issues can be a significant time sink. It's hard to debug a problem when you can't reproduce it yourself! Making builds reproducible can eliminate this class of problem and is the primary reason I think reproducible builds are useful for embedded projects. We'll explore an example of this type of problem in the next section.
+
+Having reproducible builds can also be helpful when trying to recreate the debug symbols for an image where the symbols were not preserved or to be able to re-generate a build that hasn't been updated in a while (e.g. an image used in the factory).
+
+Open source projects also strive for reproducible builds for security reasons -- to ensure no backdoors were added into the final binary as part of the compilation process[^9] [^10]
+
+## Making a Build Reproducible
+
+In this article we will examine a very simple app running FreeRTOS which passes a message between two tasks. Every time a message is sent a 64 bit counter is incremented. We will compile the binary from two different directories. If the build is reproducible, we should find the binaries match and behave the same when flashed on target. However, what we will find out is that one of the binaries crashes on boot and one does not!
+
+#### Example Project Setup
+
+In this article we will be using:
+
+- the GNU Arm Embedded Toolchain 8-2019-q3-update[^1] for our compiler
+- the nRF52840-DK[^2] (ARM Cortex-M4F) as our development board
+- SEGGER JLinkGDBServer[^3] as our GDB Server.
+
+#### Enforcing a Compiler Version
```

Comment:
Not sure why, but this doesn't get rendered as a heading:

![Screenshot 2019-12-11 at 09 39 47](https://user-images.githubusercontent.com/193881/70605137-2d0fb700-1bfa-11ea-9281-873def3712b3.png)

Reply:
hrm, interesting, what browser are you using? This is what I see:

![image](https://user-images.githubusercontent.com/7494333/70632014-1702f600-1bfc-11ea-8864-ceac7effa30a.png)

Reply:
Safari Version 13.0.3 (15608.3.10.1.4)

Reply:
Oh, wait -- I was reading the "discobot" cross-posting on the forum, not interrupt.memfault.com:

https://community.memfault.com/t/reproducible-firmware-builds-interrupt/112